src/net_processing.cpp

```diff
@@ -1998,17 +2022,27 @@ static bool ProcessGetCFHeaders(CNode* pfrom, CDataStream& vRecv, const CChainPa
         return true;
     }
 
+    bool index_in_sync = false;
```

Comment:
Both the interface and implementation of `QueryFilterIndexWithRetry` are quite complex and a bit difficult to reason about, especially when the output parameter `in_sync` is not used by the caller.

Would it be detrimental to first unconditionally call `index->BlockUntilSyncedToCurrentChain()` at each call site instead of only calling it before retrying? It would make the code more readable by eliminating the retry function and all the complexities around it (e.g. output parameter dependency).

Reply:
Yes, I agree that `QueryFilterIndexWithRetry` is quite ugly, but I really don't like the idea of blocking the net thread on background processing unless necessary. In fact, I don't like the idea of doing it at all, but there's not really infrastructure for doing proper async logic here.

Reply:
That seems like an orthogonal concern to my point. From what I can tell, `BlockUntilSyncedToCurrentChain` won't block needlessly:

https://github.com/bitcoin/bitcoin/blob/46d6930f8c7ba7cbcd7d86dd5d0117642fcbc819/src/index/base.h#L95-L100

That is, it is short-circuited if the index is current:

https://github.com/bitcoin/bitcoin/blob/b2a6b0216192b6e8428f1a80b47f5178fccb961a/src/index/base.cpp#L282-L289

So in best case you would just verify that the index is caught up before querying it. And in worse case, you would block until caught up just as you're already doing in `QueryFilterIndexWithRetry`.

Would also obliviate the need to address some of @Talkless's nits. ;)

Reply:
Yes, you are right. I guess I wanted to avoid taking cs_main, but it already does in `PrepareBlockFilterRequest` so maybe that's an unnecessary optimization.